sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py

```diff
@@ -489,75 +507,153 @@ def _resolve_inputs_for_pipeline_component_jobs(self, jobs, base_path):
-            else:
-                msg = f"Non supported job type in Pipeline: {type(job_instance)}"
-                raise ComponentException(
-                    message=msg,
-                    target=ErrorTarget.COMPONENT,
-                    no_personal_data_message=msg,
-                    error_category=ErrorCategory.USER_ERROR,
-                )
 
-    def _resolve_arm_id_and_inputs(self, component):
+            for extra_operation in extra_operations:
+                extra_operation(job_instance, key)
+
+            if isinstance(job_instance, BaseNode) and isinstance(job_instance._component, PipelineComponent):
+                if cur_layer_i + 1 == len(layers):
+                    layers.append([])
+                layers[cur_layer_i+1].extend(job_instance.component.jobs.items())
+
+            if cur_head_in_layer == len(layers[cur_layer_i]):
+                cur_layer_i += 1
+                cur_head_in_layer = 0
+
+        # if there is no subgraph, pop the empty layer inserted at the beginning
+        if len(layers[-1]) == 0:
+            layers.pop()
+
+        return layers
+
+    def _resolve_dependencies_for_pipeline_component_jobs(
+        self,
+        component: Union[Component, str],
+        resolver: Callable,
+        *,
+        resolve_inputs: bool = True
+    ):
+        """Resolve dependencies for pipeline component jobs.
+        Will directly return if component is not a pipeline component.
+
+        :param component: The pipeline component to resolve.
+        :type component: Union[Component, str]
+        :param resolver: The resolver to resolve the dependencies.
+        :type resolver: Callable
+        :param resolve_inputs: Whether to resolve inputs.
+        :type resolve_inputs: bool
+        """
         if not isinstance(component, PipelineComponent) or not component.jobs:
             return
-        self._resolve_inputs_for_pipeline_component_jobs(component.jobs, component._base_path)
-        self._resolve_arm_id_for_pipeline_component_jobs(component.jobs, self._orchestrators.get_asset_arm_id)
+
+        from azure.ai.ml.entities._job.automl.automl_job import AutoMLJob
+
+        if resolve_inputs:
+            self._resolve_inputs_for_pipeline_component_jobs(component.jobs, component._base_path)
+
+        layers = self._divide_nodes_to_resolve_into_layers(
+            component,
+            extra_operations=[
+                self._set_default_display_name_for_anonymous_component_in_node,
+            ]
+        )
+
+        # cache anonymous component only for now
+        # request level in-memory cache can be a better solution for other type of assets as they are
+        # relatively simple and of small number of distinct instances
+        component_cache = CachedNodeResolver(resolver=resolver)
+        # start from the last layer and resolve layer by layer to ensure all dependencies are resolved
+        for layer in reversed(layers):
+            for _, job_instance in layer:
```

Comment:
can we start resolving those already leaf nodes in upper layer earlier?

Reply:
I have considered so. The problem is that the benefit seems minor unless we can dynamically add nodes to the processing queue. If not, take below tree as an example: 
A
|  \  \
B C D
 / | \  
E F G 
by layer: 3 => 3
leaf first: 5 => 1
The round of concurrent component registration is the same.

For dynamically adding parent to the process queue once all its children have been resolved, the implementation can be more complicated and maybe we can consider it in further optimization.